### ISSUE TITLE
Is there a way to have a smaller "runtime only" OpenDDS docker container?

### ISSUE THREAD

**Issue by Cingulara** [objectcomputing/OpenDDS#1383]
Is there a way to have a 'runtime only' option for a container where I can compile my app within the OpenDDS container like the Dev Guide says now, but my exectuable would be in a 'runtime only' type of container? I liken this to the C# SDK container from Microsoft being huge but the runtime version you copy your compiled code into being a lot smaller of a container in size. 

The docker-compose w/ the Info server for discovery runs 3 containers. That is 3 2GB containers running where 1 is the Info server, 1 the subscriber and 1 the producer. Is there a way to shrink the containers for runtime files only that are needed? No test scripts. No extras. Just the minimum files needed so the "publisher" executable can run correctly in that container? 

I saw the https://github.com/jrw972/KubernetesMessenger info and the https://opendds.org/quickstart/GettingStartedDocker.html information and both worked well. However, the image itself is 2GB when I do a `docker images | grep opendds` and see the actual size.  I am trying to not run the "SDK" version and run the "runtime only" version of this if that makes sense. 

Is everything in /opt/OpenDDS actually required to run the application? like the ./tests/ directory? Or the PROBLEM-REPORT-FORM file? I am asking so I can shrink the image size down for memory usage, transport, etc.

**Comment by Cingulara** [comment]
I am coming from a NATS.io type of messaging so this is more robust from the server end and all you can do with it. I can compile golang clients for subscribers to a single executable and copy to a Dockerfile "FROM scratch" to make a 11MB size image as an example. That is a very simplistic example of course. 

If I look at the .NET Core 3.0 SDK image it is 1.74 GB.
If I look at the .NET Core runtime image it is 189 MB.
And look at the .NET Core ASPNET is 207 MB.

I can build in the SDK, the copy contents into one of the runtimes to make my image smaller. That is what I am getting at. Not sure if there is a way to do that here? if it is even possible with the architecture? if it is on the roadmap?

**Comment by mitza-oci** [comment]
You can build OpenDDS as static libraries (I believe this is the approach that golang uses).  Or even with shared libraries, you can copy just what you need into your container.

**Comment by Cingulara** [comment]
Ok.  Do you have any developer documentation you can show me where I should
put stuff into the container? And yes that’s exactly how Golang does it. I
need to start working on that tomorrow and I wanted to make sure I’m doing
it correctly. Or are there any pointers you can send me to on your website
to make sure I’m building that correctly?

Thank you for the quick response.
-- 
Dale Bingham
CTO and Chief Technology Evangelist
Cingulara
https://www.cingulara.com
410-984-0001

**Comment by Cingulara** [comment]
So instead of the below, I would have compiled binaries that would do the same thing that these three below do:

```
version: '2'
services:
  inforepo:
    image: objectcomputing/opendds
    command: DCPSInfoRepo -ORBListenEndpoints iiop://:3897
    volumes:
      - $PWD:/opt/workspace
  subscriber:
    image: objectcomputing/opendds
    command: ["./wait-for-it.sh", "inforepo:3897", "--", "subscriber", "-DCPSInfoRepo",  "inforepo:3897" ]
    volumes:
      - $PWD:/opt/workspace
  publisher:
    image: objectcomputing/opendds
    command: ["./wait-for-it.sh", "inforepo:3897", "--", "publisher", "-DCPSInfoRepo",  "inforepo:3897" ]
    volumes:
      - $PWD:/opt/workspace
```

**Comment by mitza-oci** [comment]
There are a few things to consider in terms of how to deploy your application:

1. Is InfoRepo discovery what you want, or would peer-to-peer discovery be better?
Note how InfoRepo discovery introduced the added complexity of the `wait-for-it` script in the above Docker example.
See the article "Bringing Multicast to the Cloud..." on https://opendds.org/about/articles/mnb.html for more details on doing peer-based discovery on cloud systems.

2. As I mentioned, OpenDDS supports static library builds.  You can enable that option at compile-time.

3. Without static libraries, you can copy just the `lib*.so` files that are loaded to the deployed system.  This is true regardless of the use of containers.

We can help you make the most effective use of OpenDDS in your application.  For timely responses please consider our commercial support option: https://objectcomputing.com/products/open-source-support

**Comment by Cingulara** [comment]
I was just talking to my cyber and system admin guy on InfoRepo versus the peer-to-peer. We will prob test both. I read that article on the cloud multicast earlier yes definitely will revisit that.  I am going to look at how to do static library builds or how to load the .so files and test that as well definitely. I will start there and see options on that. I do like I can use the opendds container to do the builds so easily. 

Thank you for such a quick reply!

**Comment by Cingulara** [comment]
inside the opendds container I ran the /opt/OpenDDS/configure script and am testing there. I will test this out to get where I need to be, and then put back in here what I did that works so we can tie a bow on it and close it.

```
./configure --target=linux --ace=ACE_wrappers --tao=/opt/OpenDDS/ACE_wrappers/TAO/ --prefix=install --static

make
```

**Comment by Cingulara** [comment]
I ran the above configure inside the container, and then the make command and then make install. I did not get my installation directory like https://github.com/objectcomputing/OpenDDS/blob/master/INSTALL.md#installation talks to unfortunately.  I also tried the --prefix with a full path to a folder and it remained empty.  I am not sure what I am doing wrong in the configure script for this not to work.

**Comment by iguessthislldo** [comment]
About your configure script usage, I would suggest dropping `--target`, `--ace` and `--tao`. The ace and tao values are the defaults and it won't download it again if you already have them. Then actually setting the target is probably a bad idea if you're not cross compiling. Also `--prefix=install` is not an absolute path. I would suggest giving it an absolute path. It might be resolving the path to `/opt/OpenDDS/install`, which is what I think you want, but I doubt that. Because of the going from cross compiling to not and the prefix change, I would make sure you have a fresh OpenDDS and ACE/TAO before you try this again. Finally I would highly recommend adding `--no-tests` unless you want to build the tests for some reason.

So to sum up, something like this:

```
rm -fr /opt/OpenDDS
# Get a completely new OpenDDS
cd /opt/OpenDDS
./configure --prefix=/opt/OpenDDS/install --static --no-tests
# Build and install
```

If `make install` is still not doing anything, please look at the exit status and the output to see if everything is working from its perspective and post what it's saying.

**Comment by Cingulara** [comment]
Ok thanks @iguessthislldo (great handle!).  I am doing all of this inside the OpenDDS container. So I just blew it away and started over. If this cannot be done inside the OpenDDS container then let me know. I am working to do this in steps in the container now, hoping to automate the CI/CD of this later to put into a known good container that is only runtime. That is the end goal. I probably should have stated that. 

When I run this inside the container I get the message below which is why I referenced it. Same with the TAO stuff. 

* ERROR: Can't find ACE at /usr/local/share/ace.

If I cannot do this from w/in a container to get a good runtime-only then I will have to try something above like the "you can copy just the lib*.so files that are loaded to the deployed system. This is true regardless of the use of containers." and find all the libxxxxx.so files required. And any variables/paths/etc. needed to make it work correctly.

**Comment by Cingulara** [comment]
Running that script inside the container I get no /opt/OpenDDS/install directory at all. Nothing seems to have changed. Running ` make install ` gives me the below output however I am not sure that does you any good.

`
root@171e0614c909:/opt/OpenDDS# make install
make[1]: Entering directory '/opt/OpenDDS/dds/idl'

GNUmakefile: /opt/OpenDDS/dds/idl/GNUmakefile.opendds_idl MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.opendds_idl | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s exe_output,installdata_files \
  -o exe_output=bin -x installdata_files -b installdata_files=share/dds   \
  /usr/local
ln -sf /usr/local/bin/opendds_idl /usr/local/share/dds/bin
make[1]: Leaving directory '/opt/OpenDDS/dds/idl'
make[1]: Entering directory '/opt/OpenDDS/dds'

GNUmakefile: /opt/OpenDDS/dds/GNUmakefile.OpenDDS_Dcps MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Dcps | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files,idl_files,typesupport_files,installdata_files \
  -b header_files=include/dds -b template_files=include/dds -b inline_files=include/dds -b idl_files=include/dds -b typesupport_files=include/dds -o lib_output=lib -x installdata_files -b installdata_files=share/dds   \
  /usr/local
echo export DDS_ROOT=/usr/local/share/dds> /usr/local/share/dds/dds-devel.sh
ln -sf /usr/local/include/dds/Version.h /usr/local/share/dds/dds
cp /opt/OpenDDS/user_macros.GNU /usr/local/share/dds 2>/dev/null || true
make[1]: Leaving directory '/opt/OpenDDS/dds'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal/GNUmakefile.MessengerMinimal_Idl MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal'
make[1]: Entering directory '/opt/OpenDDS/tools/modeling/codegen/model'

GNUmakefile: /opt/OpenDDS/tools/modeling/codegen/model/GNUmakefile.Model_Lib MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.Model_Lib | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/model -b template_files=include/model -b inline_files=include/model -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/tools/modeling/codegen/model'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/transport/tcp'

GNUmakefile: /opt/OpenDDS/dds/DCPS/transport/tcp/GNUmakefile.OpenDDS_Tcp MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Tcp | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/DCPS/transport/tcp -b template_files=include/dds/DCPS/transport/tcp -b inline_files=include/dds/DCPS/transport/tcp -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/transport/tcp'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/InfoRepoDiscovery'

GNUmakefile: /opt/OpenDDS/dds/DCPS/InfoRepoDiscovery/GNUmakefile.OpenDDS_InfoRepoDiscovery MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_InfoRepoDiscovery | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files,idl_files \
  -b header_files=include/dds/DCPS/InfoRepoDiscovery -b template_files=include/dds/DCPS/InfoRepoDiscovery -b inline_files=include/dds/DCPS/InfoRepoDiscovery -b idl_files=include/dds/DCPS/InfoRepoDiscovery -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/InfoRepoDiscovery'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal/GNUmakefile.MessengerMinimal_Publisher MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal/GNUmakefile.MessengerMinimal_Subscriber MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger.minimal'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger/GNUmakefile.Messenger_Idl MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger/GNUmakefile.Messenger_Publisher MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger/GNUmakefile.Messenger_Subscriber MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy/GNUmakefile.Messenger_ZeroCopy_Idl MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy/GNUmakefile.Messenger_ZeroCopy_Publisher MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy/GNUmakefile.Messenger_ZeroCopy_Subscriber MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/DCPS/Messenger_ZeroCopy'
make[1]: Entering directory '/opt/OpenDDS/dds/FACE'

GNUmakefile: /opt/OpenDDS/dds/FACE/GNUmakefile.OpenDDS_FACE MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_FACE | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/FACE -b template_files=include/dds/FACE -b inline_files=include/dds/FACE -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/FACE'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/FACE/Simple/GNUmakefile.FaceMessenger_Idl MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/RTPS'

GNUmakefile: /opt/OpenDDS/dds/DCPS/RTPS/GNUmakefile.OpenDDS_Rtps MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Rtps | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files,idl_files,typesupport_files \
  -b header_files=include/dds/DCPS/RTPS -b template_files=include/dds/DCPS/RTPS -b inline_files=include/dds/DCPS/RTPS -b idl_files=include/dds/DCPS/RTPS -b typesupport_files=include/dds/DCPS/RTPS -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/RTPS'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/transport/rtps_udp'

GNUmakefile: /opt/OpenDDS/dds/DCPS/transport/rtps_udp/GNUmakefile.OpenDDS_Rtps_Udp MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Rtps_Udp | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/DCPS/transport/rtps_udp -b template_files=include/dds/DCPS/transport/rtps_udp -b inline_files=include/dds/DCPS/transport/rtps_udp -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/transport/rtps_udp'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/FACE/Simple/GNUmakefile.FaceMessenger_Callbacksubscriber MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/FACE/Simple/GNUmakefile.FaceMessenger_Publisher MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'
make[1]: Entering directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'

GNUmakefile: /opt/OpenDDS/DevGuideExamples/FACE/Simple/GNUmakefile.FaceMessenger_Subscriber MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/DevGuideExamples/FACE/Simple'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/QOS_XML_Handler'

GNUmakefile: /opt/OpenDDS/dds/DCPS/QOS_XML_Handler/GNUmakefile.OpenDDS_QOS_XML_XSC_Handler MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_QOS_XML_XSC_Handler | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/DCPS/QOS_XML_Handler -b template_files=include/dds/DCPS/QOS_XML_Handler -b inline_files=include/dds/DCPS/QOS_XML_Handler -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/QOS_XML_Handler'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/security'

GNUmakefile: /opt/OpenDDS/dds/DCPS/security/GNUmakefile.OpenDDS_Security MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Security | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files,idl_files,typesupport_files \
  -b header_files=include/dds/DCPS/security -b template_files=include/dds/DCPS/security -b inline_files=include/dds/DCPS/security -b idl_files=include/dds/DCPS/security -b typesupport_files=include/dds/DCPS/security -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/security'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/transport/multicast'

GNUmakefile: /opt/OpenDDS/dds/DCPS/transport/multicast/GNUmakefile.OpenDDS_Multicast MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Multicast | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/DCPS/transport/multicast -b template_files=include/dds/DCPS/transport/multicast -b inline_files=include/dds/DCPS/transport/multicast -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/transport/multicast'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/transport/shmem'

GNUmakefile: /opt/OpenDDS/dds/DCPS/transport/shmem/GNUmakefile.OpenDDS_Shmem MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Shmem | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/DCPS/transport/shmem -b template_files=include/dds/DCPS/transport/shmem -b inline_files=include/dds/DCPS/transport/shmem -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/transport/shmem'
make[1]: Entering directory '/opt/OpenDDS/dds/DCPS/transport/udp'

GNUmakefile: /opt/OpenDDS/dds/DCPS/transport/udp/GNUmakefile.OpenDDS_Udp MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_Udp | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files \
  -b header_files=include/dds/DCPS/transport/udp -b template_files=include/dds/DCPS/transport/udp -b inline_files=include/dds/DCPS/transport/udp -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/DCPS/transport/udp'
make[1]: Entering directory '/opt/OpenDDS/dds/InfoRepo'

GNUmakefile: /opt/OpenDDS/dds/InfoRepo/GNUmakefile.DCPSInfoRepo_Lib MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.DCPSInfoRepo_Lib | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output \
   -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/InfoRepo'
make[1]: Entering directory '/opt/OpenDDS/dds/InfoRepo'

GNUmakefile: /opt/OpenDDS/dds/InfoRepo/GNUmakefile.DCPSInfoRepo_Federator MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.DCPSInfoRepo_Federator | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files,idl_files,typesupport_files \
  -b header_files=include/dds/InfoRepo -b template_files=include/dds/InfoRepo -b inline_files=include/dds/InfoRepo -b idl_files=include/dds/InfoRepo -b typesupport_files=include/dds/InfoRepo -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/InfoRepo'
make[1]: Entering directory '/opt/OpenDDS/dds/InfoRepo'

GNUmakefile: /opt/OpenDDS/dds/InfoRepo/GNUmakefile.DCPSInfoRepo_Serv MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.DCPSInfoRepo_Serv | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output \
   -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/InfoRepo'
make[1]: Entering directory '/opt/OpenDDS/dds/InfoRepo'

GNUmakefile: /opt/OpenDDS/dds/InfoRepo/GNUmakefile.DCPSInfoRepo_Main MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.DCPSInfoRepo_Main | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s exe_output \
  -o exe_output=bin    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/InfoRepo'
make[1]: Entering directory '/opt/OpenDDS/dds/monitor'

GNUmakefile: /opt/OpenDDS/dds/monitor/GNUmakefile.OpenDDS_monitor MAKEFLAGS=w

perl -ne'if (/^#END MPC-Generated Install Info/) {exit 0}' \
  -e'elsif (/^#BEGIN MPC-Generated Install Info/) {$ok=1}' \
  -e'elsif ($ok && s/^#//) {print}' GNUmakefile.OpenDDS_monitor | \
/opt/OpenDDS/ACE_wrappers/MPC/prj_install.pl -i -s lib_output,header_files,template_files,inline_files,idl_files,typesupport_files \
  -b header_files=include/dds/monitor -b template_files=include/dds/monitor -b inline_files=include/dds/monitor -b idl_files=include/dds/monitor -b typesupport_files=include/dds/monitor -o lib_output=lib    \
  /usr/local
make[1]: Leaving directory '/opt/OpenDDS/dds/monitor'
make[1]: Entering directory '/opt/OpenDDS/tools/dcpsinfo_dump'

GNUmakefile: /opt/OpenDDS/tools/dcpsinfo_dump/GNUmakefile.dcpsinfo_dump MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/tools/dcpsinfo_dump'
make[1]: Entering directory '/opt/OpenDDS/tools/repoctl'

GNUmakefile: /opt/OpenDDS/tools/repoctl/GNUmakefile.repoctl MAKEFLAGS=w

make[1]: Leaving directory '/opt/OpenDDS/tools/repoctl'
`

**Comment by iguessthislldo** [comment]
It looks like the official OpenDDS container is setting the *_ROOT variables to those locations, which the configure script is going to pick up on. So no, I would not try doing this in that container. You could probably get it to work by unsetting those variables, but you should also be able to do this in just about any other container.

**Comment by Cingulara** [comment]
Ahh these below Ok. So I need to copy down and put all the source somewhere to build this myself and then see about a runtime only once I compile in the official one I believe.

`ACE_ROOT=/usr/local/share/ace
DDS_ROOT=/usr/local/share/dds
TAO_ROOT=/usr/local/share/tao`

**Comment by Cingulara** [comment]
I have a runtime OpenDDS working for the `DCPSInfoRepo -ORBListenEndpoints iiop://:3897` as well as the publisher and subscriber demo code finally. ~ 397MB in size so not bad from shrinking from 2.1GB. I can use the larger one as my "SDK" container and compile in a pipeline. And then the "runtime" one to run after some thorough testing.

**Comment by Cingulara** [comment]
I could work to figure out from /scratch what I needed to add minimally to baseline it if I have time. But 397MB means I got it to 20% of the size. So that works for me for now. Just needed the binary and libraries for the runtime so far. I did not add openssl or anything else so I will need to for what I am doing. 

I will have a web api endpoint to call that kicks off an OpenDDS message to send out. So I wanted a minimum one to start. I really appreciate your insight and help!

**Comment by iguessthislldo** [comment]
If size is a concern, you can to configure with `--no-debug`. I think OpenDDS and friends should be in the ballpark of 10 MB in that kind of configuration, depending on what you include.

**Comment by Cingulara** [comment]
I will try to do that next. Thanks.

**Comment by mitza-oci** [comment]
```
$ file publisher
publisher: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/l, for GNU/Linux 3.2.0, BuildID[sha1]=287d963f0dfb79d4d3c48641e002e042a1877a6e, not stripped
$ size publisher
   text    data     bss     dec     hex filename
7542265  726520   12584 8281369  7e5d19 publisher
```

**Comment by Cingulara** [comment]
I got this to work with 3 separate smaller docker images with "--net=host" but not without that. So the issue now is what ports/protocols do I need to open or setup in my INI file to make this run within containers and only opening up the required stuff. So that I do not have to run all in one docker-compose or even all on one machine.

**Comment by mitza-oci** [comment]
The best way to find the ports used is with wireshark/tcpdump.  It's very dependent on your configuration, for both transport and discovery.  The OpenDDS Developer's Guide lists how to assign port numbers (and the parameters used to compute port numbers) for the various discovery/transport options.

Please note that our team here at OCI will be out of the office for the holiday weekend.

**Comment by simpsont-oci** [comment]
Also worth noting that docker has had issues with multicast traffic in it's
overlay network in the past, so traffic between containers may require
additional creative solutions (avoiding rtps discovery, rtps relay, or
using a different networking plugin for docker).

**Comment by Cingulara** [comment]
I am doing all that you said in this now.   Glad I am on the right track
thanks! I will let you know what I come up with in case others are working
this as well.
-- 
Dale Bingham
CTO and Chief Technology Evangelist
Cingulara
https://www.cingulara.com
410-984-0001
